File: python/5_Longest_Palindromic_Substring.py
class Solution:
    # @param {string} s
    # @return {string}
    def longestPalindrome(self, s):
        if not s or len(s)<2:
            return s
        news="#".join(s)
        news="#"+news+"#"
        maxlen,anchor,p,length=1,0,{},len(news)
        p[0]=1
        index=0
        for cur in range(1,length):
            if p[anchor]+anchor>cur:
                p[cur]=min(p[anchor-(cur-anchor)],p[anchor]-(cur-anchor))
            else:
                p[cur]=1
            while cur>=p[cur] and cur+p[cur]<length and news[cur+p[cur]]==news[cur-p[cur]]:
                p[cur]=p[cur]+1
            if p[anchor]+anchor<p[cur]+cur:
                anchor=cur
            if maxlen<p[anchor]:
                maxlen=p[anchor]
                index=anchor
        return s[index/2-(maxlen-1)/2:index/2+maxlen/2]

File: Java/5_Longest_Palindromic_Substring.java
import java.util.Scanner;
//an example of using Test
public class Solution extends Test{
    protected void solve(Scanner sc){
        System.out.println(longestPalindrome(sc.next()));
    }
    
    public String longestPalindrome(String s) {
        char[] cstr = Join(s, '#');
        StringBuffer sb = new StringBuffer();
        if(cstr!=null){
            int len = cstr.length;
            int longcent = 0, longleng=1, index = 0;
            int curcent = 1;
            int[] point = new int[len];
            point[0]=1;
            while(curcent<len){
                if(point[index]+index>curcent){
                    point[curcent] = Math.min(point[index - (curcent - index)], point[index] - (curcent - index));
                }else{
                    point[curcent] = 1;
                }
                while(curcent>=point[curcent]&&curcent + point[curcent]<len && cstr[curcent+point[curcent]]==cstr[curcent-point[curcent]])point[curcent]++;
                if(point[index]+index<point[curcent]+curcent){
                    index = curcent;
                }
                if(longleng<point[curcent]){
                    longleng = point[curcent];
                    longcent = curcent;
                }
                curcent++;
            }
            return s.substring(longcent/2-(longleng-1)/2, longcent/2+longleng/2);
        }else{
            return s;
        }
    }
    
    public char[] Join(String s, char deli){
        if(s==null||s.length()==0)return null;
        StringBuffer sb = new StringBuffer();
        char[] cs = s.toCharArray();
        sb.append(deli);
        for(char c:cs){
            sb.append(c);
            sb.append(deli);
        }
        return sb.toString().toCharArray();
    }
    
    public void solve(){
        sc = new Scanner(System.in);
        // below is related to concrete problems
        int T = sc.nextInt();
        sc.nextLine();//jump over the first line. 
        for(int i=1;i<=T;i++){
            solve(sc);
        }
    }
    
    public static void main(String[] args){
        Solution so = new Solution();
        so.debug();
        //so.noDebug("2");
        so.solve();
    }
}